conversion and copy of arrays which differ by axis permutations

mdavezac opened the pull request:
This PR addresses the problem of converting or copying between arrays that address essentially the same data, but with different in-memory order. In other words, when two arrays share the same axis but in different order `axisnames(A) == (:a, :b, :c)` vs `axisnames(B) == (:b, :a, :c)`.

This can be useful for instance when going from array of structures to structure of arrays idioms.

ajkeller34 commented on src/core.jl:
```diff
@@ -282,6 +282,45 @@ Base.similar(A::AxisArray, ::Type{S}, ax1::Axis, axs::Axis...) where {S} = simil
+function Base.convert(I::Type{AxisArray{TT, N, AA, AAs}},
+                      array::AxisArray{T, N, A, As}
+                     )::AxisArray{TT, N, AA, AAs} where {TT, N, AA, AAs, T, A, As}
+    perm = map(x -> findfirst(y -> typeof(y) <: x, axes(array)),
```
This line will cause conversion to fail when the second type parameter of `Axis` differs, e.g. `Base.OneTo` vs. `Base.UnitRange`:

```
julia> A = AxisArray(reshape(1:24, 2, 3, 4), Axis{:a}(1:2), Axis{:b}(1:3), Axis{:c}(1:4))
3-dimensional AxisArray{Int64,3,...} with axes:
    :a, 1:2
    :b, 1:3
    :c, 1:4
And data, a 2×3×4 Base.ReshapedArray{Int64,3,UnitRange{Int64},Tuple{}}:
[:, :, 1] =
 1  3  5
 2  4  6

[:, :, 2] =
 7   9  11
 8  10  12

[:, :, 3] =
 13  15  17
 14  16  18

[:, :, 4] =
 19  21  23
 20  22  24

julia> Adense = AxisArray(zeros(Int16, size(A)), :c, :a, :b)
3-dimensional AxisArray{Int16,3,...} with axes:
    :c, Base.OneTo(2)
    :a, Base.OneTo(3)
    :b, Base.OneTo(4)
And data, a 2×3×4 Array{Int16,3}:
[:, :, 1] =
 0  0  0
 0  0  0

[:, :, 2] =
 0  0  0
 0  0  0

[:, :, 3] =
 0  0  0
 0  0  0

[:, :, 4] =
 0  0  0
 0  0  0

julia> convert(typeof(Adense), A)
ERROR: ArgumentError: Axes do not match
Stacktrace:
 [1] convert(::Type{AxisArrays.AxisArray{Int16,3,Array{Int16,3},Tuple{AxisArrays.Axis{:c,Base.OneTo{Int64}},AxisArrays.Axis{:a,Base.OneTo{Int64}},AxisArrays.Axis{:b,Base.OneTo{Int64}}}}}, ::AxisArrays.AxisArray{Int64,3,Base.ReshapedArray{Int64,3,UnitRange{Int64},Tuple{}},Tuple{AxisArrays.Axis{:a,UnitRange{Int64}},AxisArrays.Axis{:b,UnitRange{Int64}},AxisArrays.Axis{:c,UnitRange{Int64}}}}) at /Users/ajkeller/.julia/v0.6/AxisArrays/src/core.jl:295
```

If I understand correctly what this method is intended for, it seems like it would be better to compare axis names instead of axis types. The functions `axisnames` and `AxisArrays.axisname` (unexported) may be useful for this.

mdavezac replied:
Just getting back into this after some time away. Conversions between arrays using just the name is a more interesting problem. More specifically, I now have to convert axes as well since their types may differ. I'll have to think about this a bit more, but I may need to include axis conversion functions to do this properly.

ajkeller34 commented on src/core.jl:
```diff
@@ -282,6 +282,45 @@ Base.similar(A::AxisArray, ::Type{S}, ax1::Axis, axs::Axis...) where {S} = simil
+"""
+Base.copy!(dest::AxisArray{TT, N, AA, AAs},
+           source::AxisArray{T, N, A, As}) where {TT, N, AA, AAs, T, A, As} = begin
+    perm = map(x -> findfirst(y -> y == x, axes(source)), axes(dest))
```
This isn't going to check if the axes are the same up to name and length as the docstring says. Rather, the axis scaling is also compared, which you can see here:

```
julia> Axis{:a}(1:3) == Axis{:a}(2:4)
false
```

ajkeller34 commented:
Since this touches `core.jl` maybe I'll ping @timholy for another opinion, but otherwise these seem like nice improvements. LGTM after my comments are addressed. Thanks for the PR.

mdavezac commented:
I've updated the code so that equivalent axes are identified strictly by name. If I understand things correctly, that should take care of your comments, @ajkeller34.

ajkeller34 commented on test/core.jl:
```diff
@@ -77,6 +77,54 @@ H = similar(A, Float64, 1,1,1)
+Adense = AxisArray(zeros(Int16, size(A)), axes(A))
+@test typeof(@inferred(convert(typeof(Adense), A))) == typeof(Adense)
+@test axes(convert(typeof(Adense), A)) == axes(A)
+@test all(convert(typeof(Adense), A).data .== A.data)
```
I know these are just tests, but this could be written simply and more efficiently as `@test convert(typeof(Adense), A).data == A.data`

mdavezac commented:
I think you are right about the behaviour with respect to size. `copy!`'s behaviour for standard arrays is richer than I knew. Guess I should have looked before submitting the PR 😊 

I've added an `ignore_axes` as suggested and made the efficiency improvement you pointed out.
I've also added `copy!(dest, (sub-block...), src, (sub-block...))` to the mix. However, it isn't as clear how we would want to ignore axes information this time around or whether it's needed or desirable, so I have not implemented a `ignore_axes` argument for it.

More specifically, we could ignore axis information both when (a) identifying the sub-blocks and when (b) copying. But that's already implemented as is. Or we could ignore information only in (b), but that would mean the order of the axis in indices tuples matter, and hence we would need to recover the given permutation. A bit painful and somewhat non-intuitive.

Thanks for taking the time and iterating this. As long as it converges asymptotically towards better code, it's all for the best.

mdavezac commented:
Any news?

timholy commented:
This is another example of potential confusion in treating an AxisArray as both a "standard" rectangular container and as one indexed by `Axis` objects. Overall I lean towards the behavior implemented here.

I'm wondering about the `ignore_axes` though, it seems simpler to document that the user should call `copy!(dest.data, src.data)` directly.

Any thoughts, @mbauman?

mdavezac commented:
ping?

ajkeller34 commented:
Not to leave you hanging, but at this point I think I've given what input I can, and I'll defer to others for merging. Unfortunately, everyone else is probably busy with getting Julia 0.7 ready...

iamed2 commented:
I think we should remove the `ignore_axes` argument and instead suggest in the docstring that if the caller would like to ignore the axes they should use `copy!(dest.data, src.data)`. 

I would merge after that change!
